src/pg_copy.c

```diff
@@ -0,0 +1,626 @@
+/*-------------------------------------------------------------------------
+ *
+ * src/pg_copy.c
+ *
+ * This file contains implementation of COPY utility for pg_shard
+ *
+ * Copyright (c) 2014-2015, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "c.h"
+#include "fmgr.h"
+#include "funcapi.h"
+#include "libpq-fe.h"
+#include "miscadmin.h"
+#include "plpgsql.h"
+
+#include "pg_shard.h"
+#include "pg_copy.h"
+#include "connection.h"
+#include "create_shards.h"
+#include "distribution_metadata.h"
+#include "prune_shard_list.h"
+#include "ruleutils.h"
+
+#include <stddef.h>
+#include <string.h>
+
+#include "access/heapam.h"
+#include "access/htup_details.h"
+#include "access/htup.h"
+#include "access/sdir.h"
+#include "access/tupdesc.h"
+#include "access/xact.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_class.h"
+#include "catalog/pg_type.h"
+#include "commands/extension.h"
+#include "commands/copy.h"
+#include "executor/execdesc.h"
+#include "executor/executor.h"
+#include "executor/instrument.h"
+#include "executor/tuptable.h"
+#include "lib/stringinfo.h"
+#include "nodes/execnodes.h"
+#include "nodes/makefuncs.h"
+#include "nodes/memnodes.h" /* IWYU pragma: keep */
+#include "nodes/nodeFuncs.h"
+#include "nodes/nodes.h"
+#include "nodes/params.h"
+#include "nodes/parsenodes.h"
+#include "nodes/pg_list.h"
+#include "nodes/plannodes.h"
+#include "nodes/primnodes.h"
+#include "optimizer/clauses.h"
+#include "optimizer/cost.h"
+#include "optimizer/planner.h"
+#include "optimizer/var.h"
+#include "parser/parser.h"
+#include "parser/analyze.h"
+#include "parser/parse_node.h"
+#include "parser/parsetree.h"
+#include "parser/parse_type.h"
+#include "storage/lock.h"
+#include "tcop/dest.h"
+#include "tcop/tcopprot.h"
+#include "tcop/utility.h"
+#include "tsearch/ts_locale.h"
+#include "utils/builtins.h"
+#include "utils/elog.h"
+#include "utils/errcodes.h"
+#include "utils/guc.h"
+#include "utils/lsyscache.h"
+#include "utils/palloc.h"
+#include "utils/rel.h"
+#include "utils/relcache.h"
+#include "utils/snapmgr.h"
+#include "utils/tuplestore.h"
+#include "utils/memutils.h"
+
+typedef struct { 
+	bool (*Begin)(PGconn* conn);
+	bool (*Prepare)(PGconn* conn, char const* relationName, int64 shardId);
+	bool (*CommitPrepared)(PGconn* conn, char const* relationName, int64 shardId);
+	bool (*RollbackPrepared)(PGconn* conn, char const* relationName, int64 shardId);
+	bool (*Rollback)(PGconn* conn);
+} PgCopyTransactionManager;
```

Comment:
Right now it is used only in COPY. But actually there is nothing COPY specific in it. As you can see - it is just encapsulation of two-pahse commit. It can be used in other pg_shard update statements.